## Sklearn Random Forest Regressor 
### We will peform GridSearch to find the optimal parameters to be used in the Random Forest Regressor model to predict sales. Tuning of the hyperparameters are based on the MSE value obtained

In [1]:
#Importing relevant libraries
from data import train
from sklearn.svm import SVR
import numpy as np
from multiprocessing import cpu_count
from utils import StandardizedGridSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.metrics import mean_squared_error as MSE
from sklearn.model_selection import train_test_split

In [2]:
#Separating predictors from sales column
X=train.drop('sales', axis=1)
y=train.sales

### We will first fit the training data into the default Random Forest Regressor model to see the MSE value before conducting GridSearch to find the optimal hyperparameters

In [3]:
# conducting a separate train test split to fit into default Gradient Boosting Regressor Model

X_train,X_test,y_train,y_test = train_test_split(X,y, test_size=0.2,random_state =20)

In [4]:
from sklearn.ensemble import RandomForestRegressor
RF_default = RandomForestRegressor()
RF_default.fit(X_train,y_train)

RandomForestRegressor()

In [5]:
# MSE value for Random Forest Regressor default model
RF_default_MSE=np.mean((y_test - RF_default.predict(X_test))**2)

RF_default_MSE

2497.918238345863

### We will now conduct GridSearchCV to see if a better model can be obtained

In [7]:
#Retriving saved Random Forest and Bagging Models
# Random Forest and Bagging Regressor Run 1

RF_run1=StandardizedGridSearchCV.load('models/RF3.p')
RF_run1

GridSearchCV(cv=10, estimator=RandomForestRegressor(n_estimators=50), n_jobs=-1,
             param_grid={'ccp_alpha': array([0.01, 0.02, 0.03, 0.04, 0.05, 0.06, 0.07, 0.08, 0.09, 0.1 , 0.11,
       0.12, 0.13, 0.14, 0.15, 0.16, 0.17, 0.18, 0.19, 0.2 , 0.21, 0.22,
       0.23, 0.24, 0.25, 0.26, 0.27, 0.28, 0.29, 0.3 , 0.31, 0.32, 0.33,
       0.34, 0.35, 0.36, 0.37, 0.38, 0.39, 0.4 , 0.41, 0.42, 0.43, 0.44,
       0.45, 0.46, 0.47, 0.48, 0.49,...
       0.56, 0.57, 0.58, 0.59, 0.6 , 0.61, 0.62, 0.63, 0.64, 0.65, 0.66,
       0.67, 0.68, 0.69, 0.7 , 0.71, 0.72, 0.73, 0.74, 0.75, 0.76, 0.77,
       0.78, 0.79, 0.8 , 0.81, 0.82, 0.83, 0.84, 0.85, 0.86, 0.87, 0.88,
       0.89, 0.9 , 0.91, 0.92, 0.93, 0.94, 0.95, 0.96, 0.97, 0.98, 0.99,
       1.  ]),
                         'criterion': ['mse'], 'max_depth': [10],
                         'max_features': (None, 'auto')},
             refit='neg_mean_squared_error', return_train_score=True,
             scoring=['neg_mean_squared_error', 

In [8]:
# Top 10 model results for Random Forest and Bagging Regressor Run 1

RF_run1.results.head(10)

mean_fit_time  std_fit_time  \
ccp_alpha criterion max_depth max_features                                
0.40      mse       10        auto               2.899884      0.190939   
0.26      mse       10        auto               2.864082      0.130700   
0.37      mse       10        auto               2.508874      0.259479   
0.09      mse       10        NaN                2.408103      0.255367   
0.41      mse       10        auto               2.863000      0.166757   
0.75      mse       10        NaN                2.790311      0.155244   
0.16      mse       10        NaN                2.820084      0.149843   
0.82      mse       10        auto               2.568868      0.142420   
0.34      mse       10        auto               2.511113      0.195342   
0.21      mse       10        NaN                2.914651      0.204543   

                                            mean_score_time  std_score_time  \
ccp_alpha criterion max_depth max_features                                    
0.40      mse       10        auto                 0.022561        0.002941   
0.26      mse       10        auto                 0.028877        0.008176   
0.37      mse       10        auto                 0.026801        0.012177   
0.09      mse       10        NaN                  0.023288        0.007993   
0.41      mse       10        auto                 0.022919        0.003531   
0.75      mse       10        NaN                  0.027522        0.011107   
0.16      mse       10        NaN                  0.024837        0.004359   
0.82      mse       10        auto                 0.019984        0.007035   
0.34      mse       10        auto                 0.030569        0.015252   
0.21      mse       10        NaN                  0.022568        0.005315   

                                            split0_test_neg_mean_squared_error  \
ccp_alpha criterion max_depth max_features                                       
0.40      mse       10        auto                                 -755.626536   
0.26      mse       10        auto                                 -677.526201   
0.37      mse       10        auto                                 -780.857270   
0.09      mse       10        NaN                                  -685.879815   
0.41      mse       10        auto                                 -754.359279   
0.75      mse       10        NaN                                  -691.196682   
0.16      mse       10        NaN                                  -774.392890   
0.82      mse       10        auto                                 -708.267609   
0.34      mse       10        auto                                 -697.633161   
0.21      mse       10        NaN                                  -729.137507   

                                            split1_test_neg_mean_squared_error  \
ccp_alpha criterion max_depth max_features                                       
0.40      mse       10        auto                                -1130.759937   
0.26      mse       10        auto                                -1104.905571   
0.37      mse       10        auto                                -1131.299373   
0.09      mse       10        NaN                                 -1247.368889   
0.41      mse       10        auto                                -1195.760152   
0.75      mse       10        NaN                                 -1326.688986   
0.16      mse       10        NaN                                 -1150.255681   
0.82      mse       10        auto                                -1175.909523   
0.34      mse       10        auto                                -1077.643726   
0.21      mse       10        NaN                                 -1086.686377   

                                            split2_test_neg_mean_squared_error  \
ccp_alpha criterion max_depth max_features                                       
0.40      mse       10        auto                          

In [9]:
#Printing top 10 models for Random Forest and Bagging Regressor Run 1 based on test MSE value

RF_run1.results['mean_test_neg_mean_squared_error'].head(10)

ccp_alpha  criterion  max_depth  max_features
0.40       mse        10         auto           -2367.260821
0.26       mse        10         auto           -2369.330841
0.37       mse        10         auto           -2394.247759
0.09       mse        10         NaN            -2411.766285
0.41       mse        10         auto           -2411.942395
0.75       mse        10         NaN            -2416.251007
0.16       mse        10         NaN            -2422.690418
0.82       mse        10         auto           -2435.520573
0.34       mse        10         auto           -2437.530603
0.21       mse        10         NaN            -2440.069211
Name: mean_test_neg_mean_squared_error, dtype: float64

In [10]:
#We aggreagted the train and test MSE scores of run1 to see if ccp_alpha had a significant effect on the MSE values

(-RF_run1.results[['mean_train_neg_mean_squared_error', 'mean_test_neg_mean_squared_error']]).groupby('ccp_alpha').mean()

mean_train_neg_mean_squared_error  mean_test_neg_mean_squared_error
ccp_alpha                                                                     
0.01                              415.459104                       2457.605728
0.02                              403.068408                       2531.019440
0.03                              405.120889                       2595.617216
0.04                              396.390804                       2540.406584
0.05                              385.953300                       2533.883679
...                                      ...                               ...
0.96                              424.202244                       2530.222280
0.97                              422.735572                       2700.666507
0.98                              440.051091                       2652.354658
0.99                              425.853027                       2604.285165
1.00                              428.997257                       2578.758187

[100 rows x 2 columns]

### From above, we decided to narrow the range of ccp_alpha to see if there is a effect on MSE

In [11]:
#Retriving saved Random Forest and Bagging Models
# Random Forest and Bagging Regressor Run 2

RF_run2=StandardizedGridSearchCV.load('models/RFe50.p')
RF_run2

GridSearchCV(cv=10, estimator=RandomForestRegressor(n_estimators=50), n_jobs=-1,
             param_grid={'ccp_alpha': (0.1, 1, 10), 'criterion': ('mse', 'mae'),
                         'max_depth': (9, 10, 12),
                         'max_features': (None, 'auto'),
                         'max_leaf_nodes': array([ 199,  629, 1992])},
             refit='neg_mean_squared_error', return_train_score=True,
             scoring=['neg_mean_squared_error', 'r2'], verbose=10)

In [12]:
# Top 10 model results for Random Forest and Bagging Regressor Run 2

RF_run2.results.head(10)

10        NaN          629                  2.269565   
                                           199                  2.210904   

                                                           std_fit_time  \
ccp_alpha criterion max_depth max_features max_leaf_nodes                 
1.0       mse       10        auto         199                 0.171242   
                              NaN          629                 0.091926   
                    9         auto         1992                0.091620   
                                           199                 0.082004   
                    12        NaN          199                 0.195587   
0.1       mae       10        auto         629                 0.750402   
          mse       9         auto         1992                0.137142   
1.0       mse       10        auto         629                 0.183072   
0.1       mse       10        NaN          629                 0.076147   
                                           199                 0.107924   

                                                           mean_score_time  \
ccp_alpha criterion max_depth max_features max_leaf_nodes                    
1.0       mse       10        auto         199                    0.015584   
                              NaN          629                    0.015485   
                    9         auto         1992                   0.017891   
                                           199                    0.016939   
                    12        NaN          199                    0.015372   
0.1       mae       10        auto         629                    0.013823   
          mse       9         auto         1992                   0.015513   
1.0       mse       10        auto         629                    0.014795   
0.1       mse       10        NaN          629                    0.019023   
                                           199                    0.015755   

                                                           std_score_time  \
ccp_alpha criterion max_depth max_features max_leaf_nodes                   
1.0       mse       10        auto         199                   0.001685   
                              NaN          629                   0.001876   
                    9         auto         1992                  0.009554   
                                           199                   0.002284   
                    12        NaN          199                   0.002282   
0.1       mae       10        auto         629                   0.001988   
          mse       9         auto         1992                  0.004097   
1.0       mse       10        auto         629                   0.001563   
0.1       mse       10        NaN          629                   0.005073   
                                           199                   0.002150   

                                                           split0_test_neg_mean_squared_error  \
ccp_alpha criterion max_depth max_features max_leaf_nodes                                       
1.0       mse       10        auto         199                                    -683.491642   
                              NaN          629                                    -706.258133   
                    9         auto         1992                                   -736.225889   
                                           199                                    -723.570004   
                    12        NaN          199                                    -698.066201   
0.1       mae       10        auto         629                                   -1216.399332   
          mse       9         auto         1992                                   -788.833810   
1.0       mse       10        auto         629                                    -692.011369   
0.1       mse       10        NaN          629                                    -699.843315   
                                     

In [13]:
#Printing top 10 models for Random Forest and Bagging Regressor Run 2 based on test MSE value

RF_run2.results['mean_test_neg_mean_squared_error'].head(10)

ccp_alpha  criterion  max_depth  max_features  max_leaf_nodes
1.0        mse        10         auto          199              -2428.587681
                                 NaN           629              -2454.831336
                      9          auto          1992             -2484.084729
                                               199              -2486.469285
                      12         NaN           199              -2488.630722
0.1        mae        10         auto          629              -2492.699447
           mse        9          auto          1992             -2495.357914
1.0        mse        10         auto          629              -2502.546570
0.1        mse        10         NaN           629              -2503.855931
                                               199              -2504.170200
Name: mean_test_neg_mean_squared_error, dtype: float64

### Comparing the 2 GridSearchCV models as well as the default model, we see that RF_run1 is the best model as it gives the lower test MSE score of 2367.26

In [14]:
## Printing best parameters from RF_run1

RF_run1.best_estimator_

RandomForestRegressor(ccp_alpha=0.4, max_depth=10, n_estimators=50)